File: arrays/threeNumberSum.py
def threeNumberSum(array, targetSum):

    array.sort()
    triplets = []
    for i, val in enumerate(array[:-2]):
        left = i + 1
        right = len(array) - 1
        while left < right:
            currentSum = array[i] + array[left] + array[right]
            if currentSum == targetSum:
                triplets.append([array[i], array[left], array[right]])
                left += 1
                right -= 1
            elif currentSum < targetSum:
                left += 1
            elif currentSum > targetSum:
                right -= 1
    return triplets


# Example usage
array = [12, 3, 1, 2, -6, 5, -8, 6]
targetSum = 0
print(threeNumberSum(array, targetSum))  # Output: [[-8, 2, 6], [-6, 1, 5], [1, 2, 3]]

File: arrays/threeNumberSum/threeNumberSum.js
function threeNumberSum(array, targetSum) {
  array.sort((a, b) => a - b)
  const result = []

  for(let i = 0; i < array.length; i++) {
    for(let j = i + 1; j < array.length; j++){
      for(let y = j+1; y < array.length; y++) {
        const currentIterationSum = array[i] + array[j] + array[y]
        if(currentIterationSum === targetSum)
          result.push([array[i], array[j], array[y]])
      }
    }
  }

  return result
}

threeNumberSum([12, 3, 1, 2, -6, 5, -8, 6], 0)
// Output: [[-8, 2, 6], [-6, 1, 5], [1, 2, 3]]
// Time complexity: O(n^3) 